other/dpll.py

```diff
@@ -0,0 +1,400 @@
+            else:
+                symbol = literal[:2]
+                if symbol in list(self.literals.keys()):
+                    return True
+
+        self.assign(model)
+        result = False
+        for j in self.literals.values():
+            if j is True:
+                return True
+            elif j is None:
+                return None
+        for j in self.literals.values():
+            result = result or j
+        return result
+
+
+class Formula:
+    """
+    A formula represented in CNF.
+    A formula is a set of clauses.
+    For example,
+        {{A1, A2, A3'}, {A5', A2', A1}} is ((A1 v A2 v A3') and (A5' v A2' v A1))
+
+    Create two clauses and a formula with them
+    >>> c1 = Clause(["A1", "A2'", "A3"])
+    >>> c2 = Clause(["A5'", "A2'", "A1"])
+
+    >>> f = Formula([c1, c2])
+    >>> print(f)
+    { { A1 , A2' , A3 } , { A5' , A2' , A1 } }
+    """
+
+    def __init__(self, clauses):
+        """
+        Represent the number of clauses and the clauses themselves.
+        """
+        self.clauses = [c for c in clauses]
+        self.no_of_clauses = len(self.clauses)
+
+    def __str__(self):
+        """
+        To print a formula as in CNF.
+        Variable formula holds the string representation.
+        """
+        formula = "{ "
+        for i in range(0, self.no_of_clauses):
+            formula += str(self.clauses[i]) + " "
+            if i != self.no_of_clauses - 1:
+                formula += ", "
+        formula += "}"
+
+        return formula
+
+
+def generate_clause():
+    """
+    Randomly generate a clause.
+    All literals have the name Ax, where x is an integer from 1 to 5.
+    """
+    literals = []
+    no_of_literals = random.randint(1, 5)
+    base_var = "A"
+    i = 0
+    while i < no_of_literals:
+        var_no = random.randint(1, 5)
+        var_name = base_var + str(var_no)
+        var_complement = random.randint(0, 1)
+        if var_complement == 1:
+            var_name += "'"
+        if var_name in literals:
+            i -= 1
+        else:
+            literals.append(var_name)
+        i += 1
+    clause = Clause(literals)
+    return clause
+
+
+def generate_formula():
+    """
+    Randomly generate a formula.
+    """
+    clauses = []
+    no_of_clauses = random.randint(1, 10)
+    i = 0
+    while i < no_of_clauses:
+        clause = generate_clause()
+        if clause in clauses:
+            i -= 1
+        else:
+            clauses.append(clause)
+        i += 1
+    formula = Formula(set(clauses))
+    return formula
+
+
+def generate_parameters(formula):
+    """
+    Return the clauses and symbols from a formula.
+    A symbol is the uncomplemented form of a literal.
+    For example,
+        Symbol of A3 is A3.
+        Symbol of A5' is A5.
+
+    >>> c1 = Clause(["A1", "A2'", "A3"])
+    >>> c2 = Clause(["A5'", "A2'", "A1"])
+
+    >>> f = Formula([c1, c2])
+    >>> c, s = generate_parameters(f)
+    >>> l = [str(i) for i in c]
```

Comment:
```suggestion
    >>> l = [str(i) for i in c]
    >>> lobster = [str(clause) for clause in clauses]
```
Single letter variable names make code look like it was written in the 1970s but l is the worst because it looks like I and 1.